File: YiJoonYong/programmers/118668.java
import java.util.*;
class Solution {
    List<int[]> problemList;
    int[][] timeArr;
    int maxAlp, maxCop;
    
    public int solution(int alp, int cop, int[][] problems) {
        int rLen = problems.length;
        
        // 기본 알고&코딩 공부를 추가하여 문제 리스트 생성
        problemList = new ArrayList<>(rLen+2);
        problemList.add(new int[]{0, 0, 1, 0, 1});
        problemList.add(new int[]{0, 0, 0, 1, 1});
        for (int[] problem : problems) {
            problemList.add(problem);
            
            // 풀어야하는 최대 알고&코딩 수치 저장
            maxAlp = Math.max(maxAlp, problem[0]);
            maxCop = Math.max(maxCop, problem[1]);
        }
        
        // [알고, 코딩] => 최소시간
        // 최소 시간을 저장할 배열 초기화
        timeArr = new int[Math.max(alp, maxAlp)+1][Math.max(cop, maxCop)+1];
        for (int[] line : timeArr) {
            Arrays.fill(line, Integer.MAX_VALUE);
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(arr -> arr[0]));
        pq.add(new int[]{0, alp, cop}); //[time, currAlp, currCop]
        timeArr[alp][cop] = 0;
        
        int answer = 0;
        while(!pq.isEmpty()) {
            int[] target = pq.poll();
            
            // 저장한 시간보다 클 경우 continue;
            if (timeArr[target[1]][target[2]] < target[0]) {
                continue;
            }
            
            // 최대 알고&코딩력 이상으로 능력치가 올라가면 종료
            if (target[1] >= maxAlp && target[2] >= maxCop) {
                answer = target[0];
                break;
            }
            
            // 현재 능력치에서 문제를 하나씩 해결하며 pq에 추가
            for (int[] problem : problemList) {  // [0:alp_req, 1:cop_req, 2:alp_rwd, 3:cop_rwd, 4:cost]
                
                // 문제 해결 하한선보다 현재 능력치가 낮다면 continue;
                if (target[1] < problem[0] || target[2] < problem[1]) {
                    continue;
                }
                
                // 문제 해결 후 다음 능력치 계산
                int nTime = target[0] + problem[4];
                int nAlp = Math.min(target[1] + problem[2], maxAlp);
                int nCop = Math.min(target[2] + problem[3], maxCop);
                
                // 해당 능력치에 도달하기 까지 걸린 시간이 최소라면 pq에 추가
                if (nTime < timeArr[nAlp][nCop]) {
                    timeArr[nAlp][nCop] = nTime;
                    pq.add(new int[]{nTime, nAlp, nCop});
                }
            }
        }
        
        return answer;
    }
}

File: KimHyeokjin/programmers/118668.py
def solution(alp, cop, problems):
    #IDEA. dp를 이용해서 풀것임.
    #row와 col은 각각 알고력과 코딩력이 해당 index일때 도달하는데 걸리는 시간을 의미함.
    
    #최대 알고력과 최대 코딩력을 기억하기 위해 초기화
    max_alp, max_cop = 0, 0
    for lst in problems:
        max_alp = max(max_alp, lst[0])
        max_cop = max(max_cop, lst[1])
    
    #dp배열은 최대 알고력+1, 최대 코딩력+1로 선언
    dp = [[10e9] *(max_cop+1)for _ in range(max_alp+1)]
    
    #시간 초과나서 이쪽부분 고침
    #최소 확인 위치 변경
    alp = min(alp, max_alp)
    cop = min(cop, max_cop)
    #alp와 cop밑으로는 확인하지 않을것임..!
    dp[alp][cop] = 0
    
    #알고력 해당 수준부터 최대값까지돌면서
    for i in range(alp, max_alp+1):
        # 코딩력 해당 수준부터 최대값까지 돌면서
        for j in range(cop, max_cop+1):
            #만약, 현재 알고력이 최대 알고력을 초과하지 않을때만 값을 더해줌
            if i+1<=max_alp:
                dp[i+1][j] = min(dp[i+1][j], dp[i][j]+1)
            #만약, 현재 코딩력 또한 마찬가지
            if j+1<=max_cop:
                dp[i][j+1] = min(dp[i][j+1], dp[i][j]+1)
            
            #각 문제들 별로 확인하면서
            for lst in problems:
                #만약, 알고력과 코딩력이 수준을 만족한다면
                if i>=lst[0] and j>=lst[1]:
                    #크기를 넘어가지 않는 부분에 대해서 cost값을 더해준값과 비교해줌
                    #2번 실수
                    new_alp = min(i+lst[2], max_alp)
                    new_cop = min(j+lst[3], max_cop)
                    dp[new_alp][new_cop] = min(dp[new_alp][new_cop], dp[i][j]+lst[4])
    return dp[max_alp][max_cop];
